File: port_js/org/kbinani/vsq/VibratoBPList.js
/*
 * VibratoBPList.js
 * Copyright (C) 2009-2010 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
if( org == undefined ) var org = {};
if( org.kbinani == undefined ) org.kbinani = {};
if( org.kbinani.vsq == undefined ) org.kbinani.vsq = {};
if( org.kbinani.vsq.VibratoBPList == undefined ){

    org.kbinani.vsq.VibratoBPList = function(){
        this.m_list = new Array();
        if( arguments.length == 3 ){
            var strNum = arguments[0];
            var strBPX = arguments[1];
            var strBPY = arguments[2];
            var num = 0;
            num = parseInt( strNum, 10 );
            if( isNaN( num ) ){
                num = 0;
            }
            var bpx = strBPX.split( ',' );
            var bpy = strBPY.split( ',' );
            var actNum = Math.min( num, Math.min( bpx.length, bpy.length ) );
            if ( actNum > 0 ) {
                var x = new Array( actNum );
                var y = new Array( actNum );
                for ( var i = 0; i < actNum; i++ ) {
                    x[i] = parseFloat( bpx[i] );
                    y[i] = parseInt( bpy[i] );
                }

                var len = Math.min( x.length, y.length );
                for ( var i = 0; i < len; i++ ) {
                    this.m_list.push( new org.kbinani.vsq.VibratoBPPair( x[i], y[i] ) );
                }
                this.m_list.sort( org.kbinani.vsq.VibratoBPPair.compare );
            }
        }else if( arguments.length == 2 ){
            var x = arguments[0];//float[]
            var y = arguments[1];//int[]
            var len = Math.min( x.length, y.length );
            for ( var i = 0; i < len; i++ ) {
                this.m_list.push( new org.kbinani.vsq.VibratoBPPair( x[i], y[i] ) );
            }
            this.m_list.sort( org.kbinani.vsq.VibratoBPPair.compare );
        }
    };

    org.kbinani.vsq.VibratoBPList.prototype = {
        /**
         * @param x [float]
         * @param default_value [int]
         * @return [int]
         */
        getValue : function( x, default_value ) {
            if ( this.m_list.length <= 0 ) {
                return default_value;
            }
            var index = -1;
            var size = m_list.length;
            for ( var i = 0; i < size; i++ ) {
                if ( x < this.m_list[i].X ) {
                    break;
                }
                index = i;
            }
            if ( index == -1 ) {
                return default_value;
            } else {
                return this.m_list[index].Y;
            }
        },

        /**
         * @return [object]
         */
        clone : function() {
            var ret = new org.kbinani.vsq.VibratoBPList();
            for ( var i = 0; i < m_list.length; i++ ) {
                ret.m_list.push( new org.kbinani.vsqVibratoBPPair( this.m_list[i].X, this.m_list[i].Y ) );
            }
            return ret;
        },

        /**
         * @return [int]
         */
        getCount : function() {
            return this.m_list.length;
        },

        /**
         * @param index [int]
         * @return [VibratoBPPair]
         */
        getElement : function( index ) {
            return this.m_list[index];
        },

        /**
         * @param index [int]
         * @param value [VibratoBPPair]
         * @return [void]
         */
        setElement : function( ndex, value ) {
            this.m_list[index] = value;
        },

        /**
         * @return [String]
         */
        getData : function() {
            var ret = "";
            for ( var i = 0; i < this.m_list.length; i++ ) {
                ret += (i == 0 ? "" : ",") + m_list[i].X + "=" + m_list[i].Y;
            }
            return ret;
        },

        /**
         * @param value [String]
         * @return [void]
         */
        setData : function( value ) {
            var c = this.m_list.length;
            for( var i = 0; i < c; i++ ){
                this.m_list.shift();
            }
            var spl = value.split( ',' );
            for ( var i = 0; i < spl.length; i++ ) {
                var spl2 = spl[i].split( '=' );
                if ( spl2.length < 2 ) {
                    continue;
                }
                this.m_list.push( new org.kbinani.vsq.VibratoBPPair( parseFloat( spl2[0] ), parseInt( spl2[1], 10 ) ) );
            }
        },
    };

}

File: org.kbinani.vsq/VibratoBPList.cs
/*
 * VibratoBPList.cs
 * Copyright © 2009-2011 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
#if JAVA
package org.kbinani.vsq;

import java.util.*;
import java.io.*;
import org.kbinani.*;
#else
using System;
using org.kbinani;
using org.kbinani.java.util;

namespace org.kbinani.vsq
{
    using boolean = System.Boolean;
#endif

#if JAVA
    public class VibratoBPList implements Cloneable, Serializable {
#else
    [Serializable]
    public class VibratoBPList : ICloneable
    {
#endif
        private Vector<VibratoBPPair> m_list;

        public VibratoBPList()
        {
            m_list = new Vector<VibratoBPPair>();
        }

        public VibratoBPList( String strNum, String strBPX, String strBPY )
        {
            int num = 0;
            try {
                num = PortUtil.parseInt( strNum );
            } catch ( Exception ex ) {
                PortUtil.stderr.println( "org.kbinani.vsq.VibratoBPList#.ctor; ex=" + ex );
                num = 0;
            }
            String[] bpx = PortUtil.splitString( strBPX, ',' );
            String[] bpy = PortUtil.splitString( strBPY, ',' );
            int actNum = Math.Min( num, Math.Min( bpx.Length, bpy.Length ) );
            if ( actNum > 0 ) {
                float[] x = new float[actNum];
                int[] y = new int[actNum];
                for ( int i = 0; i < actNum; i++ ) {
                    try {
                        x[i] = PortUtil.parseFloat( bpx[i] );
                        y[i] = PortUtil.parseInt( bpy[i] );
                    } catch ( Exception ex ) {
                        PortUtil.stderr.println( "org.kbinani.vsq.IconParameter#.ctor; ex=" + ex );
                    }
                }

                int len = Math.Min( x.Length, y.Length );
                m_list = new Vector<VibratoBPPair>( len );
                for ( int i = 0; i < len; i++ ) {
                    m_list.add( new VibratoBPPair( x[i], y[i] ) );
                }
                Collections.sort( m_list );
            } else {
                m_list = new Vector<VibratoBPPair>();
            }
        }

        public VibratoBPList( float[] x, int[] y )
#if JAVA
            throws NullPointerException
#endif
        {
            if ( x == null ) {
#if JAVA
                throw new NullPointerException( "x" );
#else
                throw new ArgumentNullException( "x" );
#endif
            }
            if ( y == null ) {
#if JAVA
                throw new NullPointerException( "y" );
#else
                throw new ArgumentNullException( "y" );
#endif
            }
            int len = Math.Min( x.Length, y.Length );
            m_list = new Vector<VibratoBPPair>( len );
            for ( int i = 0; i < len; i++ ) {
                m_list.add( new VibratoBPPair( x[i], y[i] ) );
            }
            Collections.sort( m_list );
        }

        /// <summary>
        /// このインスタンスと，指定したVibratoBPListのインスタンスが等しいかどうかを調べます
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public boolean equals( VibratoBPList item )
        {
            if ( item == null ) {
                return false;
            }
            int size = this.m_list.size();
            if ( size != item.m_list.size() ) {
                return false;
            }
            for ( int i = 0; i < size; i++ ) {
                VibratoBPPair p0 = this.m_list.get( i );
                VibratoBPPair p1 = item.m_list.get( i );
                if ( p0.X != p1.X ) {
                    return false;
                }
                if ( p0.Y != p1.Y ) {
                    return false;
                }
            }
            return true;
        }

        public int getValue( float x, int default_value )
        {
            if ( m_list.size() <= 0 ) {
                return default_value;
            }
            int index = -1;
            int size = m_list.size();
            for ( int i = 0; i < size; i++ ) {
                if ( x < m_list.get( i ).X ) {
                    break;
                }
                index = i;
            }
            if ( index == -1 ) {
                return default_value;
            } else {
                return m_list.get( index ).Y;
            }
        }

        public Object clone()
        {
            VibratoBPList ret = new VibratoBPList();
            for ( int i = 0; i < m_list.size(); i++ ) {
                ret.m_list.add( new VibratoBPPair( m_list.get( i ).X, m_list.get( i ).Y ) );
            }
            return ret;
        }

#if !JAVA
        public object Clone()
        {
            return clone();
        }
#endif

        public int getCount()
        {
            return m_list.size();
        }

        public VibratoBPPair getElement( int index )
        {
            return m_list.get( index );
        }

        public void setElement( int index, VibratoBPPair value )
        {
            m_list.set( index, value );
        }

#if !JAVA
        /// <summary>
        /// XMLシリアライズ用
        /// </summary>
        public String Data
        {
            get
            {
                return getData();
            }
            set
            {
                setData( value );
            }
        }
#endif

        public String getData()
        {
            String ret = "";
            for ( int i = 0; i < m_list.size(); i++ ) {
                ret += (i == 0 ? "" : ",") + m_list.get( i ).X + "=" + m_list.get( i ).Y;
            }
            return ret;
        }

        public void setData( String value )
        {
            m_list.clear();
            String[] spl = PortUtil.splitString( value, ',' );
            for ( int i = 0; i < spl.Length; i++ ) {
                String[] spl2 = PortUtil.splitString( spl[i], '=' );
                if ( spl2.Length < 2 ) {
                    continue;
                }
                m_list.add( new VibratoBPPair( PortUtil.parseFloat( spl2[0] ), PortUtil.parseInt( spl2[1] ) ) );
            }
        }
    }

#if !JAVA
}
#endif
